sycamore/transforms/merge_elements.py

```diff
@@ -0,0 +1,124 @@
+from abc import ABC, abstractmethod
+
+from ray.data import Dataset
+
+from sycamore.data import Document, Element, BoundingBox
+from sycamore.plan_nodes import NonCPUUser, NonGPUUser, Transform, Node
+from sycamore.transforms.map import generate_map_function
+from sycamore.functions.tokenizer import Tokenizer
+
+from typing import Tuple
+
+
+class ElementMerger(ABC):
+    @abstractmethod
+    def merge_elements(self, document: Document) -> Document:
+        pass
+
+
+class GreedyTextElementMerger(ElementMerger):
+    def __init__(self, tokenizer: Tokenizer, max_tokens: int):
+        self.tokenizer = tokenizer
+        self.max_tokens = max_tokens
+
+    def _should_merge(self, element1: Tuple[Element, int], element2: Tuple[Element, int]) -> bool:
+        if element2[0].type == "Title":
+            return False
+        if element1[1] + 1 + element2[1] > self.max_tokens:
+            return False
+        return True
+
+    def _merge(self, element1: Tuple[Element, int], element2: Tuple[Element, int]) -> Tuple[Element, int]:
+        """
+        Merge two elements; the new element's fields will be set as:
+
+            type: "Section"
+            binary_representation: elt1.binary_representation + elt2.binary_representation
+            text_representation: elt1.text_representation + '\n' + elt2.text_representation
+            bbox: the minimal bbox that contains both elt1's and elt2's bboxes
+            properties: elt1's properties + any of elt2's properties that are not in elt1
+                > note: if elt1 and elt2 have different values for the same property, we take elt1's value
+
+            > note: if any input field is None we take the other element's field without merge logic
+
+        Args:
+            element1 (Tuple[Element, int]): the first element (and number of tokens in it)
+            element2 (Tuple[Element, int]): the second element (and number of tokens in it)
+
+        Returns:
+            Tuple[Element, int]: a new merged element from the inputs (and number of tokens in it)
+        """
+        elt1, tok1 = element1
+        elt2, tok2 = element2
+        new_toks = 0
+        new_elt = Element()
+        new_elt.type = "Section"
+        # Merge binary representations by concatenation
+        if elt1.binary_representation is None or elt2.binary_representation is None:
+            new_elt.binary_representation = elt1.binary_representation or elt2.binary_representation
+        else:
+            new_elt.binary_representation = elt1.binary_representation + elt2.binary_representation
```

Comment:
Not sure if it matters, but I am assuming we will be able to decode it back to text if required?

Reply:
I think so, but I haven't tested all kinds of documents. This is just doing concatenation of the bytes. If the data is in a format that doesn't concatenate into something meaningful then you won't get something meaningful. Maybe we should add more knobs, i.e.
ElementMerger takes as params a merge function for each element field. So something like 
```
ElementMerger(
    binary_merge_fn,
    text_merge_fn,
    bbox_merge_fn,
    type_merge_fn,
    property_merge_fn
)
```
Could also throw in a `should_merge_fn` and make it fully general? But again, you could just say, "You want different behavior? Then implement your own subclass! You only need to write one function, it's not that hard"

Reply:
I guess this is kind of what I was getting at with my abstraction question above. I don't feel that strongly about it at this point, but it does seem like it would be nice to not have to implement a separate `GreedyElementMergerThatConcatenatesBinary` and `GreedyElementMergerThatReplacesBinary`, etc. Guess it depends on how many points in the space we expect to be useful in practice.